config/routes.rb

```diff
@@ -1,4 +1,13 @@
 Rails.application.routes.draw do
+  devise_for :users
+  as :user do
+    get 'users', :to => 'users#index', :as => :user_root
+  end
```

Comment:
ここなんですが、 [Devise の README.md](https://github.com/heartcombo/devise#controller-filters-and-helpers) の以下の記述を参考にしました。

> After signing in a user, confirming the account or updating the password, Devise will look for a scoped root path to redirect to. For instance, when using a :user resource, the user_root_path will be used if it exists; otherwise, the default root_path will be used. This means that you need to set the root inside your routes:
> 
> ```
> root to: 'home#index'
> ```

たしかに、 `as :user do ~ end` の指定を外すと、プロフィール更新後の戻り先が `user_index_path` から `book_index_path` に変わってしまったので、挙動に変化がありました。
( `root to: "books#index"` でアプリのデフォルトルートが指定され、プロフィール更新後もそこに飛んでしまっているからです)

なので、こちらの設定は残しておいてもよいかなと思っていますがいかがでしょうか。
より明示的には、devise のコントローラをカスタムしてプロフィール更新後の戻り先を変更するのも選択肢かなと思いますが、どちらがよさそうでしょうか。個人的には、devise の仕様なのでこのままでも良いかと思っていますが、 `user_index_path` の指定が `resources :users, only: [:index, :show]` と重複してしまっているところは少しスマートさに欠けるとは思っています。

Reply:
なるほど〜 👀
`user_root_path` をdeviseがチェックしているのですね。挙動が変わってしまうのでしたら、残しておきましょうか。

感想としては次のように思いました！ 🙏 

- 暗黙的なdeviseの仕様なので、コメントをつけておくとわかりやすいかも
- 一方で `root to: "books#index"` が事実上仕事をしていない？ `user_root_path` の個別定義でなく `root to: "users#index"` で済ませることも検討してもよいかも 

補足ですが、これまで自分が見てきたプロジェクトだと、リダイレクト先の明示は ApplicationController で ` after_sign_in_path_for` のoverrideを利用しているケースが多いかなと思います。メソッド名から何のために指定しているのかが分かりやすいから良く利用されているのかもしれません。( こちらばかり使っていたので `user_root_path` のやり方は知らなかったです！ :bow: )

> You can also override `after_sign_in_path_for` and `after_sign_out_path_for` to customize your redirect hooks.

一方で `user_root_path` についてもちゃんとREADMEに書いてあることなので、間違いでは無いと思います 👍